### PR TITLE
add docs and readme

### PR DESCRIPTION
**Pull request by BusyJay**
Add a simple README and fill all missing docs.

### CODE REVIEW

**Comment by siddontang** [comment]
PTAL @queenypingcap

File: README.md [hunk 1]
```diff
@@ -0,0 +1,54 @@
+# fail-rs
+
+A fail points implementation for rust.
```
Comment (queenypingcap) [inline README.md]: points -> point. 
rust -> Rust
Please also describe its function.

File: README.md [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+## Usage
+
+First, add this to your `Cargo.toml`:
```
Comment (queenypingcap) [inline README.md]: this -> the following dependency

File: README.md [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+```toml
+[dependencies]
+fail = "0.1"
+```
+
+Next, add this to your crate:
```
Comment (queenypingcap) [inline README.md]: this -> the following code

File: README.md [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+```rust
+#[macro_use]
+extern crate fail;
+```
+
+Define fail points:
```
Comment (queenypingcap) [inline README.md]: fail points -> the fail points

File: README.md [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    fail_point!("name2", |r| r.map_or(2, |e| e.parse().unwrap()));
+    0
+}
+
+fn function_conditional(enable: bool) {
+    fail_point!("name3", enable, |_| {});
+}
+```
+
+Trigger a fail point via environment variable:
```
Comment (queenypingcap) [inline README.md]: environment variable -> the environment variable

File: README.md [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+```
+$ FAILPOINTS=bar::foo=panic cargo run
+```
+
```
Comment (queenypingcap) [inline README.md]: Please Add the following:
## Test

File: README.md [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+In unit tests:
+
+```
+#[test]
+fn test_foo() {
+    fail::cfg("bar::foo", "panic");
+    foo();
+}
+```
+
```
Comment (queenypingcap) [inline README.md]: Please Add the following:
## TO DO

File: README.md [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+Triggered via HTTP API is planned but not implemented yet.
```
Comment (queenypingcap) [inline README.md]: -> Trigger a fail point via the HTTP API.

File: src/lib.rs [hunk 9]
```diff
@@ -11,13 +11,57 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-//! A failpoints implementation for rust.
+//! A fail points implementation for rust.
```
Comment (queenypingcap) [inline src/lib.rs]: fail points -> fail point

File: src/lib.rs [hunk 10]
```diff
@@ -11,13 +11,57 @@
...
+//!
+//! Fail points are code points that used to inject error by user at runtime.
```
Comment (queenypingcap) [inline src/lib.rs]: that used -> that are used
error -> errors

File: src/lib.rs [hunk 11]
```diff
@@ -11,13 +11,57 @@
...
+//! This crate is inspired by FreeBSD's [failpoints](https://freebsd.org/cgi/man.cgi?query=fail).
+//! Unlike FreeBSD's implementation, this crate only support configuration from
+//! environment variables and API.
```
Comment (queenypingcap) [inline src/lib.rs]: -> this crate only supports the configuration from environment variables and HTTP API.

File: src/lib.rs [hunk 12]
```diff
@@ -11,13 +11,57 @@
...
+//!
+//! ## Installation
+//!
+//! Add this to your `Cargo.toml`:
```
Comment (queenypingcap) [inline src/lib.rs]: Please refer to the comments in the Readme file.

File: src/lib.rs [hunk 13]
```diff
@@ -11,13 +11,57 @@
...
+//!
+//! fn main() {
+//!    fail::setup();
+//!    fail::cfg("rust_out::example", "return").unwrap();
+//!    f1();
+//!    fail::teardown();
+//! }
+//! ```
+//!
+//! The above example defines a fail point named "example". And then configure it as `return`.
```
Comment (queenypingcap) [inline src/lib.rs]: -> The above example defines a fail point named "example" and then configures it as `return`.

File: src/lib.rs [hunk 14]
```diff
@@ -11,13 +11,57 @@
...
+//! So function `f1` will return early and never panic. You can also configure it via environment
```
Comment (queenypingcap) [inline src/lib.rs]: function `f1` -> the `f1` function

File: src/lib.rs [hunk 15]
```diff
@@ -11,13 +11,57 @@
...
+//! variable `FAILPOINTS=rust_out::example=return`. For more supported configuration, please see
```
Comment (queenypingcap) [inline src/lib.rs]: You can also configure it via the `FAILPOINTS=rust_out::example=return` environment variable.

Comment (queenypingcap) [reply src/lib.rs]: -> For more supported configuration, see

File: src/lib.rs [hunk 16]
```diff
@@ -11,13 +11,57 @@
...
+//! docs for macro [`fail_point`](macro.fail_point.html) and [`setup`](fn.setup.html).
+//!
+//! If you want to disable all fail point at compile time, you can enable features `disabled`.
```
Comment (queenypingcap) [inline src/lib.rs]: If you want to disable all fail point at compile time, you can enable features `disabled`.
->
If you want to disable all the fail points during compiling, you can set the features to be "disabled".

File: src/lib.rs [hunk 17]
```diff
@@ -267,9 +308,22 @@ lazy_static! {
     static ref REGISTRY: FailPointRegistry = FailPointRegistry::default();
 }
 
+/// Setup the fail point system.
```
Comment (queenypingcap) [inline src/lib.rs]: Setup -> Set up

File: src/lib.rs [hunk 18]
```diff
@@ -267,9 +308,22 @@ lazy_static! {
...
+///
+/// The environment variable `FAILPOINTS` will be used to configure fail points.
```
Comment (queenypingcap) [inline src/lib.rs]: The `FAILPOINTS` environment variable is used to configure all the fail points.

File: src/lib.rs [hunk 19]
```diff
@@ -267,9 +308,22 @@ lazy_static! {
...
+/// `FAILPOINTS` should be in the format of `full_path_to_failpoint=actions;...`.
```
Comment (queenypingcap) [inline src/lib.rs]: The format of `FAILPOINTS` is `full_path_to_failpoint=actions;...`.

File: src/lib.rs [hunk 20]
```diff
@@ -267,9 +308,22 @@ lazy_static! {
...
+///
+/// `full_path_to_failpoint` is the full module path to the fail point plus its name.
```
Comment (queenypingcap) [inline src/lib.rs]: plus -> and

File: src/lib.rs [hunk 21]
```diff
@@ -267,9 +308,22 @@ lazy_static! {
...
+/// For more informations, please see macro [`fail_point`](macro.fail_point.html) and
```
Comment (queenypingcap) [inline src/lib.rs]: For more informations, please see
->
For more information, see

File: src/lib.rs [hunk 22]
```diff
@@ -283,6 +337,9 @@ pub fn setup() {
     }
 }
 
+/// Teardown the fail point system.
```
Comment (queenypingcap) [inline src/lib.rs]: Teardown -> Tear down

File: src/lib.rs [hunk 23]
```diff
@@ -267,9 +308,22 @@ lazy_static! {
...
+///
+/// `full_path_to_failpoint` is the full module path to the fail point plus its name.
+/// For more informations, please see macro [`fail_point`](macro.fail_point.html) and
+/// [`cfg`](fn.cfg.html).
 pub fn setup() {
     let mut registry = REGISTRY.registry.write().unwrap();
-    let failpoints = env::var("FAILPOINTS").unwrap();
+    let failpoints = match env::var("FAILPOINTS") {
+        Ok(s) => s,
+        // It's either not present or invalid unicode.
```
Comment (queenypingcap) [inline src/lib.rs]: Does this mean:
It's absent or the character is invalid Unicode?

File: src/lib.rs [hunk 24]
```diff
@@ -283,6 +337,9 @@ pub fn setup() {
     }
 }
 
+/// Teardown the fail point system.
+///
+/// All paused fail point will be notified before deactivated.
```
Comment (queenypingcap) [inline src/lib.rs]: -> All the paused fail points will be notified before they are deactivated.

File: src/lib.rs [hunk 25]
```diff
@@ -304,11 +361,36 @@ pub fn eval<R, F: FnOnce(ReturnValue) -> R>(name: &str, f: F) -> Option<R> {
     p.eval(name).map(f)
 }
 
+/// Set new actions to a fail point at runtime.
+///
+/// Actions should be in the format of `action[->action...]`. When multiple action
```
Comment (queenypingcap) [inline src/lib.rs]: The format of the actions is: `action[->action...]`. When multiple actions are specified, an action will be checked only if its former action is not triggered.

File: src/lib.rs [hunk 26]
```diff
@@ -304,11 +361,36 @@ pub fn eval<R, F: FnOnce(ReturnValue) -> R>(name: &str, f: F) -> Option<R> {
...
+/// is specified, action will be examined only when its former action is not triggered.
+///
+/// Action should be in the format `[p%][cnt*]task[(arg)]`. `p%` is the expected probability
```
Comment (queenypingcap) [inline src/lib.rs]: The format of an Action is: `[p%][cnt*]task[(arg)]`.

File: src/lib.rs [hunk 27]
```diff
@@ -304,11 +361,36 @@ pub fn eval<R, F: FnOnce(ReturnValue) -> R>(name: &str, f: F) -> Option<R> {
...
+/// action is triggered, `cnt*` is the max times the action can be triggered. Supported `task`
```
Comment (queenypingcap) [inline src/lib.rs]: action is triggered -> that an Action is triggered.

File: src/lib.rs [hunk 28]
```diff
@@ -304,11 +361,36 @@ pub fn eval<R, F: FnOnce(ReturnValue) -> R>(name: &str, f: F) -> Option<R> {
...
+/// includes:
+///
+/// - `off`, the fail point will do nothing.
+/// - `return(arg)`, return early when the fail point is triggered. `arg` is passed to `$e` (define
```
Comment (queenypingcap) [inline src/lib.rs]: define -> defined

File: src/lib.rs [hunk 29]
```diff
@@ -304,11 +361,36 @@ pub fn eval<R, F: FnOnce(ReturnValue) -> R>(name: &str, f: F) -> Option<R> {
...
+/// via `fail_point!` macro) as string.
```
Comment (queenypingcap) [inline src/lib.rs]: via `fail_point!` macro -> via the `fail_point!` macro as a string.

File: src/lib.rs [hunk 30]
```diff
@@ -304,11 +361,36 @@ pub fn eval<R, F: FnOnce(ReturnValue) -> R>(name: &str, f: F) -> Option<R> {
...
+/// - `sleep(milliseconds)`, sleep for the specified time.
+/// - `panic(msg)`, panic with the message.
+/// - `print(msg)`, print the message.
+/// - `pause`, sleep until other action is set to the fail point.
+/// - `yield`, yield the CPU.
+/// - `delay(milliseconds)`, busy waiting for the specified time.
```
Comment (queenypingcap) [inline src/lib.rs]: Sorry, I don't understand the meaning...

File: src/lib.rs [hunk 31]
```diff
@@ -304,11 +361,36 @@ pub fn eval<R, F: FnOnce(ReturnValue) -> R>(name: &str, f: F) -> Option<R> {
...
+/// message "still alive!" and 80% chance to panic. And the message will be printed at most 3
+/// times.
 pub fn cfg<S: Into<String>>(name: S, actions: &str) -> Result<(), String> {
     let mut registry = REGISTRY.registry.write().unwrap();
     set(&mut registry, name.into(), actions)
 }
 
+/// Remove a fail point.
+///
+/// If the fail point doesn't exists, nothing will happen.
```
Comment (queenypingcap) [inline src/lib.rs]: exists -> exist

File: src/lib.rs [hunk 32]
```diff
@@ -333,6 +415,30 @@ fn set(
     Ok(())
 }
 
+/// The only entry for define a fail point.
```
Comment (queenypingcap) [inline src/lib.rs]: for -> to

File: src/lib.rs [hunk 33]
```diff
@@ -333,6 +415,30 @@ fn set(
...
+///
+/// When a fail point is defined, it's reference via the full module path and name in the
```
Comment (queenypingcap) [inline src/lib.rs]: reference  -> referenced

File: src/lib.rs [hunk 34]
```diff
@@ -333,6 +415,30 @@ fn set(
...
+/// format of `crate::module::submodule::fail_point_name`. For example, library A defines
+/// a fail point in lib.rs as follow:
```
Comment (queenypingcap) [inline src/lib.rs]: follow -> follows

File: src/lib.rs [hunk 35]
```diff
@@ -333,6 +415,30 @@ fn set(
...
+///     fail_point!("p1");
+/// }
+///
+/// mod M {
+///     fn f() {
+///         fail_point!("p2");
+///     }
+/// }
+/// ```
+/// The full name of fail point `p1` is `A::p1`, and for `p2` is `A::M::p2`.
```
Comment (queenypingcap) [inline src/lib.rs]: The full name of the `p1` fail point is `A::p1`, and `p2` is `A::M::p2`.

File: src/lib.rs [hunk 36]
```diff
@@ -333,6 +415,30 @@ fn set(
...
+///
+/// `$e` is used to transform a string to the return type of outer function or closure. If
```
Comment (queenypingcap) [inline src/lib.rs]: transform -> transfer
return -> returned
outer -> external?

File: src/lib.rs [hunk 37]
```diff
@@ -333,6 +415,30 @@ fn set(
...
+/// the return type is `()`, then you can use shortcut `fail_point!($name)`.
```
Comment (queenypingcap) [inline src/lib.rs]: return -> returned
shortcut `fail_point!($name)` -> the `fail_point!($name)` shortcut.

File: src/lib.rs [hunk 38]
```diff
@@ -333,6 +415,30 @@ fn set(
...
+///
+/// If you provide an additional condition `$cond`, then the condition will be evaluated
+/// before fail point is actually examined.
```
Comment (queenypingcap) [inline src/lib.rs]: before fail point is actually examined. -> before the fail point is actually checked.

File: README.md (hunk 2 above)
Comment (BusyJay) [reply README.md]: This means this crate.

File: README.md (hunk 6 above)
Comment (BusyJay) [reply README.md]: It's not a test, it's a different usage for fail points other than the environment variables.

File: src/lib.rs (hunk 11 above)
Comment (BusyJay) [reply src/lib.rs]: Environment variable, API and HTTP API are planned to be supported. But only first two methods are implemented.

File: README.md [hunk 39]
```diff
@@ -0,0 +1,59 @@
...
+#[test]
+fn test_foo() {
+    fail::cfg("bar::foo", "panic");
+    foo();
+}
+```
+
+## To DO
+
+Triggered a fail point via the HTTP API is planned but not implemented yet.
```
Comment (queenypingcap) [inline README.md]: Triggered -> Triggering

**Comment by BusyJay** [comment]
ping

**Comment by siddontang** [comment]
LGTM

**Comment by siddontang** [comment]
PTAL @queenypingcap

**Comment by queenypingcap** [comment]
LGTM

**Comment by siddontang** [comment]
CI failed @BusyJay

**Pull request merged**